google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/StreamWriter.java

```diff
@@ -0,0 +1,856 @@
+            .setTransportChannelProvider(builder.channelProvider)
+            .setEndpoint(builder.endpoint)
+            .build();
+    shutdown = new AtomicBoolean(false);
+    refreshAppend();
+  }
+
+  /** Stream name we are writing to. */
+  public String getStreamNameString() {
+    return streamName;
+  }
+
+  /**
+   * Schedules the writing of a message. The write of the message may occur immediately or be
+   * delayed based on the writer batching options.
+   *
+   * <p>Example of writing a message.
+   *
+   * <pre>{@code
+   * AppendRowsRequest message;
+   * ApiFuture<AppendRowsResponse> messageIdFuture = writer.append(message);
+   * ApiFutures.addCallback(messageIdFuture, new ApiFutureCallback<AppendRowsResponse>() {
+   *   public void onSuccess(AppendRowsResponse response) {
+   *     if (response.hasOffset()) {
+   *       System.out.println("written with offset: " + response.getOffset());
+   *     } else {
+   *       System.out.println("received an in stream error: " + response.error().toString());
+   *     }
+   *   }
+   *
+   *   public void onFailure(Throwable t) {
+   *     System.out.println("failed to write: " + t);
+   *   }
+   * }, MoreExecutors.directExecutor());
+   * }</pre>
+   *
+   * @param message the message in serialized format to write to BigQuery.
+   * @return the message ID wrapped in a future.
+   */
+  public ApiFuture<AppendRowsResponse> append(AppendRowsRequest message) {
+    Preconditions.checkState(!shutdown.get(), "Cannot append on a shut-down writer.");
+
+    final AppendRequestAndFutureResponse outstandingAppend =
+        new AppendRequestAndFutureResponse(message);
+    List<InflightBatch> batchesToSend;
+    messagesBatchLock.lock();
+    try {
+      batchesToSend = messagesBatch.add(outstandingAppend);
+      // Setup the next duration based delivery alarm if there are messages batched.
+      setupAlarm();
+      if (!batchesToSend.isEmpty()) {
+        for (final InflightBatch batch : batchesToSend) {
+          LOG.fine("Scheduling a batch for immediate sending.");
+          writeBatch(batch);
+        }
+      }
+    } finally {
+      messagesBatchLock.unlock();
+    }
+
+    return outstandingAppend.appendResult;
+  }
+
+  public void refreshAppend() throws IOException {
+    synchronized (this) {
+      Preconditions.checkState(!shutdown.get(), "Cannot append on a shut-down writer.");
+      if (stub != null) {
+        stub.shutdown();
+      }
+      backgroundResourceList.remove(stub);
+      stub = BigQueryWriteClient.create(stubSettings);
+      backgroundResourceList.add(stub);
+      backgroundResources = new BackgroundResourceAggregation(backgroundResourceList);
+      messagesBatch.resetAttachSchema();
+      bidiStreamingCallable = stub.appendRowsCallable();
+      clientStream = bidiStreamingCallable.splitCall(responseObserver);
+      try {
+        while (!clientStream.isSendReady()) {
+          Thread.sleep(100);
```

Comment:
You're doing this while synchronized -- is this a good practice?

Reply:
This is to wait for the channel to become ready. Is there any good practice suggestions here?

Reply:
Does this wait have to happen within the synchronized block? since the thread needs to be locked every time for 100msc to wait -- limiting the scope of the synchronize block could help?

Reply:
Best be, since if the stream is not ready, it cannot be used to send. I will reduce the wait time to 5ms.

Reply:
SG